### PR TITLE
Introduce transfer_weights and StrictLoad.KEY_MATCHING mode

### PR DESCRIPTION
**Pull request by BloodAxe**
This PR adds `StrictLoad.KEY_MATCHING` mode to do partial checkpoint load. This is different from `.load_state_dict(state_dict, strict=False)` in one but critical part - if the layer has matching name, but shape is different, we will just skip this layer and load what has matching shape, ignoring the rest. Vanilla `load_state_dict` would just raise an error.

See https://github.com/Deci-AI/super-gradients/blob/09c0c0d6ca966a84fc320566fa410fc233770bae/tests/unit_tests/load_checkpoint_test.py#L10 test to get the idea what it does.

The motivation for this change is to have an ability to do partial checkpoint load. In case of DEKR, the pretrained weights comping from another repo and NO_KEY_MATCHING load strategy cannot do the work. `NO_KEY_MATCHING` doing index-based loading and it puts very strict requirement on the order of layers stored in checkpoint (side note: IMO it is very, very, very wroong. We had chat with @shaydeci about this and he explained this is to support some old checkpoints. ).
With `KEY_MATCHING` we could use these pretrained checkpoints without issues.

### CODE REVIEW

File: src/super_gradients/training/utils/checkpoint_utils.py [hunk 1]
```diff
@@ -172,6 +197,9 @@ def load_checkpoint_to_model(
     @param load_weights_only:
     @return:
     """
+    if isinstance(strict, str):
```
Comment (shaydeci) [inline src/super_gradients/training/utils/checkpoint_utils.py:200]: Thanks for being a good citizen (:

**Review by shaydeci — APPROVED** [review]
LGTM

File: src/super_gradients/training/utils/checkpoint_utils.py (hunk 1 above)
Comment (BloodAxe) [reply src/super_gradients/training/utils/checkpoint_utils.py:200]: 🤗
